core/embed/extmod/modtrezorutils/modtrezorutils.c

```diff
@@ -117,6 +119,49 @@ STATIC mp_obj_t mod_trezorutils_halt(size_t n_args, const mp_obj_t *args) {
 STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_trezorutils_halt_obj, 0, 1,
                                            mod_trezorutils_halt);
 
+/// def firmware_hash(challenge: bytes | None) -> bytes:
+///     """
+///     Computes the Blake2s hash of the firmware with an optional challenge as
+///     the key.
+///     """
+STATIC mp_obj_t mod_trezorutils_firmware_hash(size_t n_args,
+                                              const mp_obj_t *args) {
+  BLAKE2S_CTX ctx;
+  mp_buffer_info_t chal = {0};
+  if (n_args > 0 && args[0] != mp_const_none) {
+    mp_get_buffer_raise(args[0], &chal, MP_BUFFER_READ);
+  }
+
+  if (chal.len != 0) {
+    if (blake2s_InitKey(&ctx, BLAKE2S_DIGEST_LENGTH, chal.buf, chal.len) != 0) {
+      mp_raise_msg(&mp_type_ValueError, "Invalid challenge.");
+    }
+  } else {
+    blake2s_Init(&ctx, BLAKE2S_DIGEST_LENGTH);
+  }
+
+  for (int i = 0; i < FIRMWARE_SECTORS_COUNT; i++) {
+    uint8_t sector = FIRMWARE_SECTORS[i];
```

Comment:
It seems this counts the unused spaces too? Is that a problem?

Am I correct that this relies on user storage being located before firmware? It would be nice to assert that somehow.

Reply:
> It seems this counts the unused spaces too?

If you mean the unused space at the end of the firmware, then yes. The client that is checking the firmware needs to take that into account and pad the hashed data with `0xff` bytes, see `tools/firmware_hash.py`. The alternative would be to parse the vendor header and the firmware header to obtain the total size of the firmware. That's not too much trouble I think, but it might prove to be a pointless exercise if we go ahead with padding the firmware with random data.

Reply:
> Am I correct that this relies on user storage being located before firmware?

I am not sure, you mean because of the MPU, so that it's readable? Shouldn't it always be readable?

Reply:
No, I misunderstood how `flash_sector_size()` calculates the size and thought that any holes in the sector space would also be counted. So the whole comment is moot.

That opens a different angle however: we actually **should** be counting the unused regions (sectors 12-16 on TT), otherwise that's 128 kB of completely hidden space.

Speaking of, the user storage is also 128 kB total hidden space. No idea what to do about that.

Reply:
> That opens a different angle however: we actually should be counting the unused regions (sectors 12-16 on TT), otherwise that's 128 kB of completely hidden space.

Yes, ideally we should. But in order for it to make sense we would have to fill it with (known) random data. I'd also put these regions after the firmware in the hashing process, so that it's easier to compute on the client.

> Speaking of, the user storage is also 128 kB total hidden space. No idea what to do about that.

Me neither at the moment.